lib/str-repr/decode.c

```diff
@@ -0,0 +1,212 @@
+      control = ch;
+      break;
+    }
+  g_string_append_c(value, control);
+}
+
+static inline gboolean
+_invoke_match_delimiter(const gchar *cur, const gchar **new_cur, const StrReprDecodeOptions *options)
+{
+  if (options->delimiter_chars[0])
+    {
+      if (*cur == options->delimiter_chars[0] || *cur == options->delimiter_chars[1])
+        {
+          if (options->match_delimiter)
+            return options->match_delimiter(cur, new_cur, options->match_delimiter_data);
+          else
+            {
+              *new_cur = cur + 1;
+              return TRUE;
+            }
+        }
+    }
+  else
+    {
+      if (options->match_delimiter)
+        return options->match_delimiter(cur, new_cur, options->match_delimiter_data);
+    }
+  return FALSE;
+}
+
+gboolean
+str_repr_decode_append_with_options(GString *value, const gchar *input, const gchar **end,
+                                    const StrReprDecodeOptions *options)
+{
+  const gchar *cur = input, *new_cur;
+  gchar quote_char;
+  gint quote_state;
+  gsize initial_len = value->len;
+
+  quote_state = KV_QUOTE_INITIAL;
+  while (*cur)
+    {
+      switch (quote_state)
+        {
+        case KV_QUOTE_INITIAL:
+          if (_invoke_match_delimiter(cur, &new_cur, options))
+            {
+              cur = new_cur;
+              goto finish;
+            }
+          else if (*cur == '\"' || *cur == '\'')
+            {
+              quote_state = KV_QUOTE_STRING;
+              quote_char = *cur;
+            }
+          else
+            {
+              g_string_append_c(value, *cur);
+              quote_state = KV_UNQUOTED_CHARACTERS;
+            }
+          break;
+        case KV_QUOTE_STRING:
+          if (*cur == quote_char)
+            quote_state = KV_EXPECT_DELIMITER;
+          else if (*cur == '\\')
+            quote_state = KV_QUOTE_BACKSLASH;
+          else
+            g_string_append_c(value, *cur);
+          break;
+        case KV_QUOTE_BACKSLASH:
+          _decode_backslash_escape(value, quote_char, *cur);
+          quote_state = KV_QUOTE_STRING;
+          break;
+        case KV_EXPECT_DELIMITER:
+          if (_invoke_match_delimiter(cur, &new_cur, options))
+            {
+              cur = new_cur;
+              quote_state = KV_QUOTE_INITIAL;
+              goto finish;
+            }
+          else
+            {
+              quote_state = KV_QUOTE_ERROR;
+            }
+          break;
+        case KV_QUOTE_ERROR:
+          if (_invoke_match_delimiter(cur, &new_cur, options))
+            {
+              cur = new_cur;
+              goto finish;
+            }
+          break;
+        case KV_UNQUOTED_CHARACTERS:
+          if (_invoke_match_delimiter(cur, &new_cur, options))
+            {
+              cur = new_cur;
+              goto finish;
+            }
+          else
+            {
+              g_string_append_c(value, *cur);
+            }
+          break;
+        }
```

Comment:
I have just addressed this in a subsequent patch. Note that I've refactored the function in about 6-8 steps, from which I had 6 separate commits, but I've folded them up, I hope it's ok.

Tests didn't change.